File: src/bind/point_types.cpp
#include <pcl/point_types.h>

#include <emscripten/bind.h>

using namespace pcl;
using namespace emscripten;

EMSCRIPTEN_BINDINGS(point_types) {
  value_object<PointXY>("PointXY").field("x", &PointXY::x).field("y", &PointXY::y);

  value_object<PointXYZ>("PointXYZ")
      .field("x", &PointXYZ::x)
      .field("y", &PointXYZ::y)
      .field("z", &PointXYZ::z);

  value_object<PointXYZI>("PointXYZI")
      .field("x", &PointXYZI::x)
      .field("y", &PointXYZI::y)
      .field("z", &PointXYZI::z)
      .field("intensity", &PointXYZI::intensity);

  value_object<InterestPoint>("InterestPoint")
      .field("x", &InterestPoint::x)
      .field("y", &InterestPoint::y)
      .field("z", &InterestPoint::z)
      .field("strength", &InterestPoint::strength);

  value_object<PointXYZL>("PointXYZL")
      .field("x", &PointXYZL::x)
      .field("y", &PointXYZL::y)
      .field("z", &PointXYZL::z)
      .field("label", &PointXYZL::label);

  value_object<PointXYZRGB>("PointXYZRGB")
      .field("x", &PointXYZRGB::x)
      .field("y", &PointXYZRGB::y)
      .field("z", &PointXYZRGB::z)
      .field("r", &PointXYZRGB::r)
      .field("g", &PointXYZRGB::g)
      .field("b", &PointXYZRGB::b);

  value_object<PointXYZRGBA>("PointXYZRGBA")
      .field("x", &PointXYZRGBA::x)
      .field("y", &PointXYZRGBA::y)
      .field("z", &PointXYZRGBA::z)
      .field("r", &PointXYZRGBA::r)
      .field("g", &PointXYZRGBA::g)
      .field("b", &PointXYZRGBA::b)
      .field("a", &PointXYZRGBA::a);

  value_object<PointXYZRGBL>("PointXYZRGBL")
      .field("x", &PointXYZRGBL::x)
      .field("y", &PointXYZRGBL::y)
      .field("z", &PointXYZRGBL::z)
      .field("r", &PointXYZRGBL::r)
      .field("g", &PointXYZRGBL::g)
      .field("b", &PointXYZRGBL::b)
      .field("label", &PointXYZRGBL::label);

  value_object<Normal>("Normal")
      .field("normal_x", &Normal::normal_x)
      .field("normal_y", &Normal::normal_y)
      .field("normal_z", &Normal::normal_z)
      .field("curvature", &Normal::curvature);

  value_object<PointNormal>("PointNormal")
      .field("x", &PointNormal::x)
      .field("y", &PointNormal::y)
      .field("z", &PointNormal::z)
      .field("normal_x", &PointNormal::normal_x)
      .field("normal_y", &PointNormal::normal_y)
      .field("normal_z", &PointNormal::normal_z)
      .field("curvature", &PointNormal::curvature);

  value_object<PointXYZLNormal>("PointXYZLNormal")
      .field("x", &PointXYZLNormal::x)
      .field("y", &PointXYZLNormal::y)
      .field("z", &PointXYZLNormal::z)
      .field("label", &PointXYZLNormal::label)
      .field("normal_x", &PointXYZLNormal::normal_x)
      .field("normal_y", &PointXYZLNormal::normal_y)
      .field("normal_z", &PointXYZLNormal::normal_z)
      .field("curvature", &PointXYZLNormal::curvature);

  value_object<PointXYZINormal>("PointXYZINormal")
      .field("x", &PointXYZINormal::x)
      .field("y", &PointXYZINormal::y)
      .field("z", &PointXYZINormal::z)
      .field("intensity", &PointXYZINormal::intensity)
      .field("normal_x", &PointXYZINormal::normal_x)
      .field("normal_y", &PointXYZINormal::normal_y)
      .field("normal_z", &PointXYZINormal::normal_z)
      .field("curvature", &PointXYZINormal::curvature);

  value_object<PointXYZRGBNormal>("PointXYZRGBNormal")
      .field("x", &PointXYZRGBNormal::x)
      .field("y", &PointXYZRGBNormal::y)
      .field("z", &PointXYZRGBNormal::z)
      .field("r", &PointXYZRGBNormal::r)
      .field("g", &PointXYZRGBNormal::g)
      .field("b", &PointXYZRGBNormal::b)
      .field("normal_x", &PointXYZRGBNormal::normal_x)
      .field("normal_y", &PointXYZRGBNormal::normal_y)
      .field("normal_z", &PointXYZRGBNormal::normal_z)
      .field("curvature", &PointXYZRGBNormal::curvature);

  value_object<PointSurfel>("PointSurfel")
      .field("x", &PointSurfel::x)
      .field("y", &PointSurfel::y)
      .field("z", &PointSurfel::z)
      .field("normal_x", &PointSurfel::normal_x)
      .field("normal_y", &PointSurfel::normal_y)
      .field("normal_z", &PointSurfel::normal_z)
      .field("r", &PointSurfel::r)
      .field("g", &PointSurfel::g)
      .field("b", &PointSurfel::b)
      .field("a", &PointSurfel::a)
      .field("radius", &PointSurfel::radius)
      .field("confidence", &PointSurfel::confidence)
      .field("curvature", &PointSurfel::curvature);
}

File: src/modules/common/point-types.ts
export class PointXY {
  // HACK: To distinguish between different types
  private readonly _PointXY = 'PointXY';
  constructor(public x = 0, public y = 0) {}
}

export class PointXYZ {
  private readonly _PointXYZ = 'PointXYZ';
  constructor(public x = 0, public y = 0, public z = 0) {}
}

export class PointXYZI {
  private readonly _PointXYZI = 'PointXYZI';

  constructor(public x = 0, public y = 0, public z = 0, public intensity = 0) {}
}

export class PointXYZL {
  private readonly _PointXYZL = 'PointXYZL';
  constructor(public x = 0, public y = 0, public z = 0, public label = 0) {}
}

export class InterestPoint {
  private readonly _InterestPoint = 'InterestPoint';
  constructor(public x = 0, public y = 0, public z = 0, public strength = 0) {}
}

export class PointXYZRGB {
  private readonly _PointXYZRGB = 'PointXYZRGB';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public r = 0,
    public g = 0,
    public b = 0,
  ) {}
}

export class PointXYZRGBA {
  private readonly _PointXYZRGBA = 'PointXYZRGBA';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public r = 0,
    public g = 0,
    public b = 0,
    public a = 0,
  ) {}
}

export class PointXYZRGBL {
  private readonly _PointXYZRGBL = 'PointXYZRGBL';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public r = 0,
    public g = 0,
    public b = 0,
    public label = 0,
  ) {}
}

export class Normal {
  private readonly _Normal = 'Normal';
  constructor(
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public curvature = 0,
  ) {}
}

export class PointNormal {
  private readonly _PointNormal = 'PointNormal';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public curvature = 0,
  ) {}
}

export class PointXYZLNormal {
  private readonly _PointXYZLNormal = 'PointXYZLNormal';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public label = 0,
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public curvature = 0,
  ) {}
}

export class PointXYZINormal {
  private readonly _PointXYZINormal = 'PointXYZINormal';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public intensity = 0,
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public curvature = 0,
  ) {}
}

export class PointXYZRGBNormal {
  private readonly _PointXYZRGBNormal = 'PointXYZRGBNormal';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public r = 0,
    public g = 0,
    public b = 0,
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public curvature = 0,
  ) {}
}

export class PointSurfel {
  private readonly _PointSurfel = 'PointSurfel';
  constructor(
    public x = 0,
    public y = 0,
    public z = 0,
    public normalX = 0,
    public normalY = 0,
    public normalZ = 0,
    public r = 0,
    public g = 0,
    public b = 0,
    public a = 0,
    public radius = 0,
    public confidence = 0,
    public curvature = 0,
  ) {}
}

// Define all point types
export type PointTypes =
  | PointXY
  | PointXYZ
  | PointXYZI
  | InterestPoint
  | PointXYZL
  | PointXYZRGB
  | PointXYZRGBA
  | PointXYZRGBL
  | Normal
  | PointNormal
  | PointXYZRGBNormal
  | PointXYZINormal
  | PointXYZLNormal
  | PointSurfel;

export type PointTypesTypeof =
  | typeof PointXY
  | typeof PointXYZ
  | typeof PointXYZI
  | typeof InterestPoint
  | typeof PointXYZL
  | typeof PointXYZRGB
  | typeof PointXYZRGBA
  | typeof PointXYZRGBL
  | typeof Normal
  | typeof PointNormal
  | typeof PointXYZRGBNormal
  | typeof PointXYZINormal
  | typeof PointXYZLNormal
  | typeof PointSurfel;

// Define all point types that include XYZ data
export type XYZPointTypes =
  | PointXYZ
  | PointXYZI
  | PointXYZL
  | PointXYZRGBA
  | PointXYZRGB
  | PointXYZRGBL
  | InterestPoint
  | PointNormal
  | PointXYZRGBNormal
  | PointXYZINormal
  | PointXYZLNormal
  | PointSurfel;

export type XYZPointTypesTypeof =
  | typeof PointXYZ
  | typeof PointXYZI
  | typeof PointXYZL
  | typeof PointXYZRGBA
  | typeof PointXYZRGB
  | typeof PointXYZRGBL
  | typeof InterestPoint
  | typeof PointNormal
  | typeof PointXYZRGBNormal
  | typeof PointXYZINormal
  | typeof PointXYZLNormal
  | typeof PointSurfel;

// Define all point types with XYZ and label
export type XYZLPointTypes = PointXYZL | PointXYZRGBL | PointXYZLNormal;

// Define all point types that include RGB data
export type RGBPointTypes =
  | PointXYZRGB
  | PointXYZRGBA
  | PointXYZRGBL
  | PointXYZRGBNormal
  | PointSurfel;

// Define all point types that include normal data
export type NormalPointTypes =
  | Normal
  | PointNormal
  | PointXYZRGBNormal
  | PointXYZINormal
  | PointXYZLNormal
  | PointSurfel;

export const pointTypeMap = {
  PointXY,
  PointXYZ,
  PointXYZI,
  InterestPoint,
  PointXYZL,
  PointXYZRGB,
  PointXYZRGBA,
  PointXYZRGBL,
  Normal,
  PointNormal,
  PointXYZRGBNormal,
  PointXYZINormal,
  PointXYZLNormal,
  PointSurfel,
};

export abstract class NativeObject {
  abstract _native: Emscripten.NativeAPI;

  public clone() {
    return this._native.clone();
  }

  public delete() {
    return this._native.delete();
  }

  public deleteLater() {
    return this._native.deleteLater();
  }

  public isDeleted() {
    return this._native.isDeleted();
  }
}

export class Vector<T> extends NativeObject {
  constructor(public _native: Emscripten.NativeAPI) {
    super();
  }

  get size(): number {
    return this._native.size();
  }

  public set(index: number, value: T): boolean {
    return this._native.set(index, value);
  }

  public get(index: number): T {
    return this._native.get(index);
  }

  public push(value: T) {
    this._native.push_back(value);
  }

  public isEmpty(): boolean {
    return this._native.empty();
  }

  public resize(count: number, value?: T) {
    this._native.resize(count, value ?? null);
  }

  public clear() {
    this._native.clear();
  }
}
